### PR TITLE
fix(prime_agent, develop): rule delivery + develop discipline + AgentDB cleanup

### PR DESCRIPTION
**Pull request by elijahr**
## Summary

Three independent fixes that all touch the prime-agent + develop-skill surface, bundled because they were requested together and split would leave the gate ledger half-implemented.

1. **Prime Agent rule delivery via session-start extension** -- the install gap that left prime-agent users with no auto-loaded behavioral context after the monolithic AGENTS.spellbook.md skill was removed in `eb62832c`. Now every selected rule module is auto-injected into the system prompt at session start, via a TypeScript extension Prime Agent discovers from `~/.prime/agent/extensions/`. The user's own `~/.prime/agent/AGENTS.md` is never touched.

2. **Develop skill discipline** -- the three fixes called out in the nmg2-prime-agent handoff:
   - Surface-clarity in `commands/feature-config.md` 0.4 that parallelization is dispatch count, not ceremony.
   - Wave-discipline enforcement: refuse "Wave X done" without a recorded check.
   - `rules/40-develop-discipline.md` rule module documents both, so an agent encountering them mid-run sees them at session start.

3. **Gate ledger state file pattern + AgentDB cleanup**:
   - `scripts/develop_gate_ledger.py` is the runtime the develop skill has referenced 12+ times but never had. Deep-merge writes, refuses to rewrite `ceremony.locked_at`, refuses `failed` wave-discipline without `open_rows`.
   - Five AgentDB references in `skills/fact-checking/SKILL.md` and seven in `commands/fact-check-verify.md` (including a TypeScript code block calling a tool that does not exist) replaced with the existing `.fact-checking/state.json` checkpoint mechanism.

## Commits

```
deb7cc9b chore(docs): regenerate docs site from skills/commands/rules
35f17dc7 fix(develop): gate ledger state file + AgentDB cleanup + discipline docs
584457c9 chore(install): update prime_agent post-install note for rules extension
0e230564 fix(develop): wave-completion discipline + parallelism/ceremony surface clarity
158bb594 fix(prime_agent): auto-inject rule modules via session-start extension
```

## Why Prime Agent needs an extension

Prime Agent auto-loads `~/.prime/agent/AGENTS.md` at session start. The user owns that file -- writing it would clobber their context. Prime Agent's documented `SYSTEM.md` / `APPEND_SYSTEM.md` mechanisms are referenced in `docs/usage.md` but the loaders are not implemented in `packages/coding-agent/src/core/system-prompt.ts` on `main` as of this writing. Prime Agent has no native `rules/` directory walker. The extension system IS implemented: Prime Agent auto-discovers `~/.prime/agent/extensions/*.ts` at session start, hooks `session_start` to pre-load data, and exposes `before_agent_start` for prompt mutation. The extension uses that mechanism to deliver every selected rule module as content (not as a list the agent has to fetch on demand).

## Why the gate ledger is a real PR change

`scripts/develop_gate_ledger.py` adds the Python reference implementation for the persistent state file the develop skill has been telling the LLM to write since at least the lock design. Before this PR, the skill said:

> develop records its own phase/gate progress in a persistent state file so the work survives context compaction and a resumed session can re-assert the remaining gates instead of declaring "done" prematurely.

There was no `develop_gate_ledger.py` and no state file. The skill was aspirational. This PR implements the contract in Python so:
- Deep-merge writes happen (the skill's "MERGE-ONLY, NEVER overwrite" rule is enforced in code, not just in prompt).
- `ceremony.locked_at` cannot be rewritten after lock.
- Wave-discipline `status=failed` cannot be written without `open_rows`.
- The skill now points at the Python CLI for the actual write operation, with explicit "DO NOT hand-write the file" warning.

## Why AgentDB removal is real

AgentDB is named five times in `skills/fact-checking/SKILL.md` and seven times in `commands/fact-check-verify.md`. It is not implemented anywhere in the spellbook runtime. The invariant principle "ALWAYS use AgentDB" is broken behavior -- the skill is instructing the LLM to use a tool that does not exist. The replacement (`.fact-checking/state.json`) is the existing run checkpoint, already implemented and already used for interruption handling. Cross-run knowledge is intentionally out of scope.

## Test status

- 25 new gate ledger tests pass
- 9 prime-agent rules tests pass
- 348 total tests in `tests/scripts/`, `tests/integration/`, `tests/installer/`, `tests/test_multi_target_*`, `tests/test_installer_config_helpers.py`, `tests/test_cli/` pass
- `uv run install.py --dry-run --platforms prime_agent` succeeds with all 5 install steps OK
- 4 pre-existing failures on `main` HEAD are unrelated and fail the same way without any of my changes

## Out of scope

- **Goose's `~/.agents/AGENTS.md` overwrite.** Goose's installer writes a symlink at `~/.agents/AGENTS.md` pointing at `rules/00-core.md`, which would clobber a user's existing file. Same bug class as prime-agent's would have been, different platform. Worth a separate PR.
- **Real-develop-run validation.** The develop skill's use of the gate ledger is not unit-tested in this PR. The Python contract is. An actual develop run against a wave-structured plan would validate the LLM-side discipline end-to-end. Deferred because it requires operator-driven execution, not a Python test.

## Checklist

- [x] Tests pass locally
- [x] `uv run install.py --dry-run --platforms prime_agent` succeeds
- [x] Ruff clean on new files
- [x] Docs regenerated from skills/commands/rules
- [x] CHANGELOG entry for prime-agent rule delivery

### CODE REVIEW

File: scripts/develop_gate_ledger.py [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+                print(
+                    f"error: unknown ceremony field {name!r}; "
+                    f"valid: {sorted(CEREMONY_FIELDS)}",
+                    file=sys.stderr,
+                )
+                return 2
+            write_ledger({"ceremony": {name: args.value}}, path=Path(args.path) if args.path else None)
+        else:
+            write_ledger({args.field: args.value}, path=Path(args.path) if args.path else None)
+    except LedgerError as exc:
```
[the remark this answers is not in the document]
Comment (elijahr) [reply scripts/develop_gate_ledger.py:359]: Fixed in 68517fb1. `_cmd_set` now routes `ceremony.*` writes through `set_ceremony_field()` rather than `write_ledger()`, so the CLI hits the same `locked_at` rewrite guard the library does.

You were right that the gap was untested — added three CLI tests: first-set succeeds, rewrite is refused with exit 1 **and the on-disk value stays at the original lock**, and an unknown `ceremony.*` field still exits 2. Asserting the on-disk value matters here; an exit code alone would not prove the write was actually prevented.

File: tests/scripts/test_develop_gate_ledger.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+sys.path.insert(0, str(SCRIPT_PATH.parent))
+
+import develop_gate_ledger as ledger
+
+
+@pytest.fixture
+def tmp_ledger(tmp_path, monkeypatch):
+    """Point the ledger at a temp file so tests do not touch the real one."""
+    target = tmp_path / "develop_gate_ledger.json"
+    monkeypatch.setattr(ledger, "DEFAULT_LEDGER_PATH", target)
```
[the remark this answers is not in the document]
Comment (elijahr) [reply tests/scripts/test_develop_gate_ledger.py]: Fixed in 68517fb1, using your suggested approach. The `tmp_ledger` fixture now sets `SPELLBOOK_DEV_DIR` via `monkeypatch.setenv` instead of patching the module constant.

This simplified more than the finding asked for: the same env var also redirects the CLI subprocesses, so the six per-test `monkeypatch.setenv` calls collapsed into the fixture.

File: tests/scripts/test_develop_gate_ledger.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+
+def _run_cli(*args: str) -> subprocess.CompletedProcess:
+    return subprocess.run(
+        [sys.executable, str(SCRIPT_PATH), *args],
+        capture_output=True,
+        text=True,
+        timeout=30,
+        check=False,
+    )
```
[the remark this answers is not in the document]
Comment (elijahr) [reply tests/scripts/test_develop_gate_ledger.py:279]: Fixed in 68517fb1 — `@pytest.mark.allow("subprocess")` is now on each CLI test.

Worth noting for future runs: your own analysis was right that the firewall `allow` list in `pyproject.toml` already permits `subprocess:*` globally, so this was convention-only, not a behavior change. Applied it anyway since AGENTS.md documents the marker.

File: scripts/develop_gate_ledger.py [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+
+def _is_scalar(v: Any) -> bool:
+    """A scalar in ledger terms: not a dict, not a list.
+
+    Strings, ints, bools, None are scalars. The skill stores
+    newline-joined SCALAR strings for things like ``remaining_gates``;
+    a list would accumulate forever under merge semantics.
+    """
+    return isinstance(v, (str, int, bool, float, type(None)))
+
```
[the remark this answers is not in the document]
Comment (elijahr) [reply scripts/develop_gate_ledger.py]: Fixed in 68517fb1 — removed. Took the delete rather than the wire-it-in option: the scalar-replacement contract is already stated in the `_deep_merge` docstring and the module docstring, and a runtime assert would add a failure mode without a caller that needs it.

File: extensions/prime-agent/spellbook-rules.ts [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+		if (!/^\d{2}-spellbook-[a-z][a-z0-9-]*\.md$/.test(name)) {
+			continue;
+		}
+
+		let text: string;
+		try {
+			text = fs.readFileSync(fullPath, "utf8");
+		} catch (err) {
+			const msg = err instanceof Error ? err.message : String(err);
+			return { rules: [], error: `cannot read ${name}: ${msg}` };
```
[the remark this answers is not in the document]
Comment (elijahr) [reply extensions/prime-agent/spellbook-rules.ts]: Fixed in ff818f4e. You were right, and the blast radius was worse than one rule: because the early return discarded `rules` wholesale, a stale symlink stripped every rule *already parsed* too — the agent came up with no behavioural context at all and no way to know it was missing. Since these are symlinks into a spellbook checkout, one going stale is ordinary rather than exceptional.

Took your suggested shape (marker rule + `continue`) with two additions:

- The marker says the guidance is **NOT loaded** and not to assume the rule’s constraints are absent. Skipping quietly would be indistinguishable from the rule never having been installed, and the agent cannot report what it never saw.
- The per-file failure no longer sets the directory-level `error`, which means “the whole load failed”. A single bad file must not claim that, or callers cannot tell the two cases apart.

`loadRules` is now exported and covered by 8 vitest cases in `tests/unit/spellbook-rules.test.ts` — the extension had no test harness before, though `tests/unit` already runs vitest in CI, so this needed no new infrastructure.

I mutation-checked rather than trusting green: reverting `loadRules` to the old early-return fails 6 of the 8 new cases. The file was restored byte-identical afterwards.

File: scripts/develop_gate_ledger.py [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+    path.parent.mkdir(parents=True, exist_ok=True)
+
+
+def _deep_merge(base: Any, overlay: Any) -> Any:
+    """Deep-merge ``overlay`` onto ``base`` per the ledger merge contract.
+
+    - dict + dict: per-key merge.
+    - scalar + scalar: overlay wins (replacement).
+    - list + list: overlay wins (replacement; the ledger does not use lists).
+    - type mismatch: overlay wins (caller is asserting a new shape).
```
[the remark this answers is not in the document]
Comment (elijahr) [reply scripts/develop_gate_ledger.py]: Fixed in ff818f4e — `_deep_merge` now warns, naming the full dotted path and the fields being discarded.

I scoped the warning to the case that destroys **structure** rather than a single value: an object being replaced by a non-object. Your `ceremony: {...}` → scalar example is exactly it, and it is the sharp one, because collapsing that object drops `locked_at` — whose entire purpose is to be un-rewritable.

Scalar-to-scalar replacement stays silent deliberately. It is the documented contract and happens on nearly every write (`current_phase`, `remaining_gates`), so warning there would train the reader to ignore the warning that matters.

The write still proceeds rather than raising: refusing it would strand a ledger that a genuine shape change had already moved past. The goal was to stop the loss being invisible, not to block it.

3 new tests cover the warning, the deliberate silence on scalar replacement, and the dotted path (`waves.3a.section_24_6_check`).

File: hooks/spellbook_hook.py [hunk 7]
```diff
@@ -384,35 +384,182 @@ def _handle_user_prompt_submit(data: dict) -> list[str]:
...
+
+    When the state lookup is broken (most common cause: daemon
+    unreachable), the caller emits ``POST_COMPACT_FALLBACK_DIRECTIVE``
+    so the user gets SOMETHING rather than a silent failure mode that
+    erases the post-compaction recovery prompt entirely.
+    """
+    try:
+        result = _mcp_call(
+            "workflow_state_load",
+            {"project_path": cwd, "max_age_hours": 24},
```
[the remark this answers is not in the document]
Comment (elijahr) [reply hooks/spellbook_hook.py]: Fixed in cfccfea6 — the whole rich branch is removed.

I verified the claim before acting, and the root cause is deeper than an unregistered tool: **there is no implementation to register.** `workflow_state_load` is absent from the server's 23 registered tools, and `load_workflow_state`, `session_init`, and `resume_boot_prompt` appear nowhere in `spellbook/` — both `spellbook/resume.py` and `spellbook/sessions/resume.py` are gone. The state subsystem was removed wholesale.

So the call could only ever raise, and what the code called a "fallback" was in truth the only path that has ever run.

The deciding argument was this PR's own: it removes AgentDB from the fact-checking skill precisely because instructing the LLM to use a tool that does not exist is broken behavior. The hook was doing the same thing, and should not be an exception to an argument the PR makes about itself.

Rebuilding the state subsystem to make the rich path real is far larger than this PR, so it was not attempted.

File: hooks/spellbook_hook.py [hunk 8]
```diff
@@ -384,35 +384,182 @@ def _handle_user_prompt_submit(data: dict) -> list[str]:
...
+    if not result or not result.get("found"):
+        return None
+
+    state = result.get("state") or {}
+    parts = [
+        "## POST-COMPACTION RECOVERY DIRECTIVE",
+        "",
+        "**CRITICAL**: Context was just compacted. You MUST take these actions IMMEDIATELY, before ANY other work:",
+        "",
+        "1. Call `spellbook_session_init` MCP tool",
```
[the remark this answers is not in the document]
Comment (elijahr) [reply hooks/spellbook_hook.py]: Fixed in cfccfea6, same removal as BOT-C2. `spellbook_session_init` is equally nonexistent, and the directive that named it is gone along with the rest of the unreachable branch.

File: hooks/spellbook_hook.py [hunk 9]
```diff
@@ -384,35 +384,182 @@ def _handle_user_prompt_submit(data: dict) -> list[str]:
...
+    output ordering.
+    """
+    fragments: list[str] = []
+
+    source = (data.get("source") or "").strip()
+
+    if source == "compact":
+        cwd = (data.get("cwd") or "").strip()
+        if cwd:
+            try:
```
[the remark this answers is not in the document]
Comment (elijahr) [reply hooks/spellbook_hook.py]: Fixed in cfccfea6 — resolved by construction rather than by adding a guard and a test for it.

The directive carries no per-project content, so the `cwd` gate had nothing to protect. It is gone entirely. That inverts the edge case you found: a compaction with a blank or absent cwd is exactly when the LLM most needs telling its context was truncated, and the old gate dropped the notice precisely then.

Covered by a new test over the missing / empty / whitespace-only cwd variants, plus one pinning that a non-compact source still emits nothing — so removing the gate cannot leak the directive into ordinary startups.

File: scripts/generate_docs.py [hunk 10]
```diff
@@ -46,6 +47,26 @@
...
+    """Return the shortest fence that can legally wrap ``body``.
+
+    Every generator below embeds a raw source body inside a fenced block so
+    that XML-style tags (``<CRITICAL>``, ``<RULE>``) are shown rather than
+    swallowed as HTML. CommonMark only requires the wrapping fence to be
+    LONGER than the longest fence inside it, so the fence length is a
+    property of the body, not a constant.
+
+    This used to be a hardcoded ten backticks everywhere. That is legal but
+    reads as noise -- a body containing no fences at all (agents/MODEL_ROUTING.md)
```
[the remark this answers is not in the document]
Comment (elijahr) [reply scripts/generate_docs.py:63]: Fixed in 6343e59e — 22 tests in `tests/scripts/test_generate_docs_fences.py`.

Worth stating why this one deserved real tests rather than a token pass: the guarantee `fence_for` owes is **one-directional**. A fence that is too long is merely noisy — that is exactly what the old hardcoded ten backticks was. A fence that is too short silently truncates the page at its first inner fence, and the generator writes 200+ files in a single pass, so an off-by-one corrupts the docs site wholesale instead of failing visibly.

So the suite asserts the invariant directly (`result is always strictly longer than the longest inner run`) rather than only pinning known input/output pairs, and the final case runs the **real corpus** — every file in `rules/`, `agents/`, and `commands/` — since those are the inputs that would actually corrupt the site.

Also covered: the floor for fence-free and empty bodies, inline code spans not counting as fences, escalation across widths, longest-run-wins rather than last-run-wins, and CommonMark's up-to-three-spaces indented fence.

Mutation-checked rather than trusted green: dropping the `+1` fails 11 of the 22, including the real-corpus case. `generate_docs.py` was restored byte-identical afterwards.

File: skills/develop/SKILL.md [hunk 11]
```diff
@@ -1028,6 +1032,69 @@ Direct/Lightweight Path (zero flags — develop STAYS RESIDENT, never exits):
...
+material progress to exactly this class of error — "Wave 3a done"
+markings made without §24.6 verification, then propagated across handoffs
+because no later step re-checked.
+
+**If the plan has no wave structure** (single flat list of tasks, no
+`Wave N:` headers, no `W<n>-` row prefixes), this check is N/A — record
+`section_24_6_check: { "status": "n_a", "reason": "plan has no wave
+structure" }` in `develop_gate_ledger.waves` (with `<wave_id>` being the
+literal string `"plan"`) so the absence of the check is itself visible.
+
```
[the remark this answers is not in the document]
Comment (elijahr) [reply skills/develop/SKILL.md:1095]: Fixed in d7583254 — implemented rather than deleted from the prose.

You offered both resolutions. I took the implement side because the field earns its place on `n_a`: the stated purpose of recording `n_a` at all is that a later reader can distinguish "the operator established this check does not apply here" from "nobody ran it". A bare `n_a` only carries the first half. Deleting the prose would have removed intent rather than a defect.

It is deliberately inert as evidence, which was the thing to get right: a `failed` entry still requires `open_rows`, and a reason neither satisfies that requirement nor makes a failed wave claimable. Narrative must never substitute for evidence. Blank strings are dropped too — an empty reason looks answered, which is worse than absent.

All three surfaces now describe one contract: the data-structure definition in `skills/develop/SKILL.md`, `record_wave_discipline()`, and a `--reason` CLI flag with a usage example.

6 new tests, one of which runs the documented CLI invocation verbatim — prose describing a flag that does not exist is precisely what you caught. Mutation-checked: dropping the store fails 2 of them, including the CLI case.
